src/abbrev.rs

```diff
@@ -0,0 +1,761 @@
+    fn empty() -> Abbreviations {
+        Abbreviations { abbrevs: hash_map::HashMap::new() }
+    }
+
+    /// Insert an abbreviation into the set.
+    ///
+    /// Returns `Ok` if it is the first abbreviation in the set with its code,
+    /// `Err` if the code is a duplicate and there already exists an
+    /// abbreviation in the set with the given abbreviation's code.
+    fn insert(&mut self, abbrev: Abbreviation) -> Result<(), ()> {
+        match self.abbrevs.entry(abbrev.code) {
+            hash_map::Entry::Occupied(_) => Err(()),
+            hash_map::Entry::Vacant(entry) => {
+                entry.insert(abbrev);
+                Ok(())
+            }
+        }
+    }
+
+    /// Get the abbreviation associated with the given code.
+    #[inline]
+    pub fn get(&self, code: u64) -> Option<&Abbreviation> {
+        self.abbrevs.get(&code)
+    }
+
+    /// Parse a series of abbreviations, terminated by a null abbreviation.
+    fn parse(mut input: &[u8]) -> ParseResult<(&[u8], Abbreviations)> {
+        let mut abbrevs = Abbreviations::empty();
+
+        loop {
+            let result = Abbreviation::parse_null(input).map(|(rest, _)| (rest, None));
+            let result =
+                result.or_else(|_| Abbreviation::parse(input).map(|(rest, a)| (rest, Some(a))));
+            let (rest, abbrev) = try!(result);
+            input = rest;
+
+            match abbrev {
+                None => break,
+                Some(abbrev) => {
+                    if let Err(_) = abbrevs.insert(abbrev) {
+                        return Err(Error::DuplicateAbbreviationCode);
+                    }
+                }
+            }
+        }
+
+        Ok((input, abbrevs))
+    }
+}
+
+/// An abbreviation describes the shape of a `DebuggingInformationEntry`'s type:
+/// its code, tag type, whether it has children, and its set of attributes.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Abbreviation {
+    code: u64,
+    tag: constants::DwTag,
+    has_children: constants::DwChildren,
+    attributes: Vec<AttributeSpecification>,
+}
+
+impl Abbreviation {
+    /// Construct a new `Abbreviation`.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `code` is `0`.
+    pub fn new(code: u64,
+               tag: constants::DwTag,
+               has_children: constants::DwChildren,
+               attributes: Vec<AttributeSpecification>)
+               -> Abbreviation {
+        assert!(code != 0);
+        Abbreviation {
+            code: code,
+            tag: tag,
+            has_children: has_children,
+            attributes: attributes,
+        }
+    }
+
+    /// Get this abbreviation's code.
+    #[inline]
+    pub fn code(&self) -> u64 {
+        self.code
+    }
+
+    /// Get this abbreviation's tag.
+    #[inline]
+    pub fn tag(&self) -> constants::DwTag {
+        self.tag
+    }
+
+    /// Return true if this abbreviation's type has children, false otherwise.
+    #[inline]
```

Comment:
Yes, the annotations on name()/form() were required to due to the switch to using them in parser.rs, and others improved the improved the benchmark. I think a micro benchmark for a single attribute would help highlight these.